code/player/TTTPlayer.VoiceChat.cs

```diff
@@ -0,0 +1,177 @@
+using System.Collections.Generic;
+
+using Sandbox;
+
+using TTTReborn.Teams;
+
+namespace TTTReborn.Player
+{
+    public partial class TTTPlayer
+    {
+        public bool IsTeamVoiceChatEnabled { get; private set; } = false;
+
+        // clientside-only
+        public bool IsSpeaking { get; internal set; } = false;
+
+        private static Dictionary<TTTPlayer, List<Client>> _oldReceiveClients = new();
+
+        private TimeSince _teamChatButtonPressPending = 0f;
+
+        private const float _pressDelayFix = 0.5f;
+
+        public void TickPlayerVoiceChat()
+        {
+            using (Prediction.Off())
+            {
+                IsSpeaking = false;
+
+                if (Input.Down(InputButton.Voice) || IsTeamVoiceChatEnabled)
+                {
+                    IsSpeaking = true;
+
+                    UI.VoiceList.Current?.OnVoicePlayed(GetClientOwner(), 1f);
+                }
+
+                if (Input.Pressed(InputButton.Run) && CanUseTeamVoiceChat(this))
+                {
+                    ConsoleSystem.Run("requestteamchat", true);
+                }
+                else if (Input.Released(InputButton.Run) && IsTeamVoiceChatEnabled)
+                {
+                    ConsoleSystem.Run("requestteamchat", false);
+                }
+
+                // Edge-case fix (if `requestteamchat false` was sent, but IsTeamVoiceChatEnabled was not already set)
+                if (Input.Down(InputButton.Run))
+                {
+                    _teamChatButtonPressPending = 0f;
+                }
+                else if (IsTeamVoiceChatEnabled && _teamChatButtonPressPending > _pressDelayFix)
+                {
+                    ConsoleSystem.Run("requestteamchat", false);
+                }
+            }
+        }
+
+        [ServerCmd(Name = "requestteamchat")]
+        public static void RequestTeamChat(bool toggle)
+        {
+            TTTPlayer player = ConsoleSystem.Caller.Pawn as TTTPlayer;
+
+            if (!player.IsValid() || toggle && !CanUseTeamVoiceChat(player))
+            {
+                return;
+            }
+
+            ToggleTeamChat(player, toggle);
+        }
+
+        private static void ToggleTeamChat(TTTPlayer player, bool toggle)
+        {
+            player.IsTeamVoiceChatEnabled = toggle;
+
+            List<Client> clients = new();
+
+            if (toggle)
+            {
+                foreach (Client client in Client.All)
+                {
+                    if (client.Pawn is TTTPlayer pawnPlayer)
+                    {
+                        if (player.Team == pawnPlayer.Team)
+                        {
+                            clients.Add(client);
+                        }
+                    }
+                }
+
+                _oldReceiveClients[player] = clients;
+            }
+            else
+            {
+                // player did not talked before
```

Comment:
Just minor grammar, `Player did not talk before`, or `Player has not talked before`